network-mux/src/Network/Mux.hs

```diff
@@ -210,23 +230,34 @@ muxChannel pmss mid md cnt = do
                 callStack
         atomically $ modifyTVar cnt (+ 1)
         atomically $ putTMVar w encoding
+        traceWith tracer $ MuxTraceChannelSendStart mid encoding
         atomically $ writeTBQueue (tsrQueue pmss) (TLSRDemand mid md want)
+        traceWith tracer $ MuxTraceChannelSendEnd mid
 
     recv :: m (Maybe BL.ByteString)
     recv = do
         -- We receive CBOR encoded messages as ByteStrings (possibly partial) from the
         -- matching ingress queueu. This is the same queue the 'demux' thread writes to.
+        traceWith tracer $ MuxTraceChannelRecvStart mid
         blob <- atomically $ do
             let q = ingressQueue (dispatchTable pmss) mid md
             blob <- readTVar q
             if blob == BL.empty
                 then retry
                 else writeTVar q BL.empty >> return blob
         -- say $ printf "recv mid %s mode %s blob len %d" (show mid) (show md) (BL.length blob)
+        traceWith tracer $ MuxTraceChannelRecvEnd mid blob
         return $ Just blob
 
 muxBearerSetState :: (MonadSTM m, Ord ptcl, Enum ptcl, Bounded ptcl)
-                  => MuxBearer ptcl m
+                  => Tracer m (MuxTrace ptcl)
+                  -> MuxBearer ptcl m
                   -> MuxBearerState
                   -> m ()
-muxBearerSetState bearer newState = atomically $ writeTVar (state bearer) newState
+muxBearerSetState tracer bearer newState = do
+    oldState <- atomically $ do
+        -- If MonadSTM had swapTVar we could use it here.
```

Comment:
Maybe turn this into a todo, I created an issue for it: https://github.com/input-output-hk/ouroboros-network/issues/1063

Reply:
Will address.